packages/components/src/components/pin-code/pin-code.tsx

```diff
@@ -0,0 +1,305 @@
+              ref={(el) => this.inputElements.push(el)}
+            />
+          ))}
+        </div>
+        {hasMessage(this.host, this.message, this.state) && (
+          <StateMessage
+            id={stateMessageId}
+            state={this.state}
+            message={this.message}
+            theme={this.theme}
+            host={this.host}
+          />
+        )}
+      </Host>
+    );
+  }
+
+  private onInput = (
+    e: InputEvent & {
+      target: HTMLInputElement;
+    }
+  ): void => {
+    // needed to update value on auto-complete via keyboard suggestion
+    const { target } = e;
+    if (target.value.length >= this.length) {
+      const sanitisedValue = removeWhiteSpaces(getSanitisedValue(this.host, target.value, this.length));
+      this.value = sanitisedValue;
+      this.emitUpdateEvent();
+      this.focusFirstEmptyOrLastInput(sanitisedValue);
+    }
+  };
+
+  private onKeyDown = (
+    e: KeyboardEvent & {
+      target: HTMLInputElement & { previousElementSibling: HTMLInputElement; nextElementSibling: HTMLInputElement };
+    }
+  ): void => {
+    const {
+      key,
+      target,
+      target: { previousElementSibling, nextElementSibling },
+    } = e;
+    // prevent default for disabled or loading, but do not impede tab key
+    if (isDisabledOrLoading(this.disabled, this.loading) && key !== 'Tab') {
+      e.preventDefault();
+    } // if input is valid overwrite old value
+    else if (isInputSingleDigit(key)) {
+      e.preventDefault();
+      target.value = key;
+      this.value = getInputValue(this.inputElements);
+      this.emitUpdateEvent();
+
+      nextElementSibling?.focus();
+    } // handle alphanumeric keys, allow copy/paste shortcut
+    else if (key.length === 1 && !(e.ctrlKey || e.metaKey)) {
+      e.preventDefault();
+    } // handle backspace and delete
+    else if (key === 'Backspace' || key === 'Delete') {
+      // transfer focus backward/forward, if the input value is empty
+      if (!target.value) {
+        e.preventDefault();
+        if (key === 'Backspace' && previousElementSibling) {
+          previousElementSibling.value = '';
+          previousElementSibling.focus();
+        } else if (key === 'Delete' && nextElementSibling) {
+          nextElementSibling.value = '';
+          nextElementSibling.focus();
+        }
+      }
+      target.value = '';
+      this.inputElements.forEach((pinCodeElement) => {
+        if (pinCodeElement === e.target) {
+          pinCodeElement.value = '';
+        }
+      });
+
+      this.value = getInputValue(this.inputElements);
+      this.emitUpdateEvent();
+    } // support native submit behavior
+    else if (key === 'Enter') {
+      if (
+        isWithinForm &&
+        (this.form.querySelectorAll('input').length === 1 || // other sibling form elements e.g. select, textarea do not prevent a submit
+          this.form.querySelector('p-button[type=submit]') ||
+          this.form.querySelector('p-button-pure[type=submit]') ||
```

Comment:
What if components are prefixed?
And what if `type` attribute is not set, but instead property is used?